keps/sig-cloud-provider/20190729-building-without-in-tree-providers.md

```diff
@@ -0,0 +1,260 @@
+/ variables / ...
+
+- A few locations in the code do not properly use the cloud provider interface,
+some of these must be updated with both a "with provider" version and a
+"without provider" version broken out of the existing code. In particular this 
+includes the in-tree volumes until CSI migration is standard, and the GCE IPAM
+logic in kube-controller-manager.
+
+In particular this adds tags / constraints to:
+- `staging/src/k8s.io/legacy-cloud-providers/*` (constraints on all the providers)
+- `pkg/cloudprovider` (constraints on importing the providers)
+- `pkg/volumes/*`, `cmd/kubelet` (versions with and without the imported providers for in-tree volumes)
+- `pkg/controller/nodeipam`, `cmd/kube-apiserver`, `cmd/kube-controller-manager` (with and without GCE IPAM)
+
+`test/*` is punted to a future follow up, and credential providers are punted
+to [the external credential provider KEP](https://github.com/kubernetes/enhancements/pull/1137).
+
+### Risks and Mitigations
+
+This is only developer facing, however we will need to ensure that these tags
+stay up to date if we want this build mode to continue to work (the normal
+build mode should work by default without any additional maintenance).
+
+To ensure this continues to work we can mitigate by:
+
+- verify in CI that the cloud provider packages have boilerplate including the
+build constraints
+- build in this mode in CI to ensure that the build succeeds
+
+## Design Details
+
+### Test Plan
+
+We will add CI to ensure that we can build with this mode enabled.
+
+Additionally, we can add CI to ensure that clusters can actually be started in
+this mode.
+
+Initially, [kind](https://github.com/kubernetes-sigs/kind) can be used to ensure
+that Kubernetes works without the providers, in the future we can extend this
+CI to out-of-tree providers combined with this build mode as their CI is spun up.
+
+### Graduation Criteria
+
+##### Alpha -> Beta Graduation
+
+Likely unnecessary, as we will eventually remove the in-tree provider code entirely for [removing-in-tree-providers](keps/sig-cloud-provider/20190125-removing-in-tree-providers.md).
+This is also not a user facing change.
+
+##### Beta -> GA Graduation
+
+Likely unnecessary, as we will eventually remove the in-tree provider code entirely for  [removing-in-tree-providers](keps/sig-cloud-provider/20190125-removing-in-tree-providers.md).
+This is also not a user facing change.
+
+Final graduation can be considered to be when the cloud provider code is actually
+removed from the Kubernetes source tree, at which point this work will be complete.
+
+### Upgrade / Downgrade Strategy
+
+N/A ?
+
+### Version Skew Strategy
+
+N/A ?
+
+## Implementation History
+
+- original prototype [kubernetes/kubernetes#80353](https://github.com/kubernetes/kubernetes/pull/80353)
+- original KEP PR [kubernetes/enhancements#1180](https://github.com/kubernetes/enhancements/pull/1180)
+
+## Drawbacks [optional]
```

Comment:
we don't also "exclude the actual vendored cloud provider code" from the build right? IIUC there's no real great way to verify (besides by hand) that all references to the cloud provider code have been removed by this build constraint.

Specifically if somehow new files get added or things get shuffled without our knowledge a dependency could sneak back in?

Reply:
It is true that I could import an arbitrary go library that happens to depend on say, google cloud specific functionality. I don't think that's a terribly feasible issue to tackle (what packages are cloud specific?) however we have a set of known cloud provider libraries in the main repo that we intend to remove in the future.

Those packages are covered in the linked KEP https://github.com/kubernetes/enhancements/blob/master/keps/sig-cloud-provider/20190125-removing-in-tree-providers.md

This KEP covers "deleting" them early with the build tag.

Reply:
Can we `go list` with the build tags and look for e.g. `aws-sdk-go` or `google-cloud-go`?

Reply:
We will only be simulating the removal of our own code (the legacy providers) to gain confidence for finishing that KEP. 

When we do remove the legacy providers it would be good to clean up these sort of cloud specific dependencies, but that's a follow on task and should probably be an update to or follow on topic for the removal KEP.

When we do go to do that cleanup however, `go mod why github.com/googleapis/google-cloud-go` should be helpful.

Reply:
I thought `go list` honored build tags, but ... agree this can come later!  Certainly easy enough to try once the build tags are added!

Reply:
It would be nice if we could piggy back on this work to prevent new cloud provider dependencies being added in tree.